### DISCUSSION TITLE
d34 model (~$2,500)

### DISCUSSION THREAD

**Discussion by karpathy** [karpathy/nanochat#314] [Announcements]
I trained and uploaded the d34 model to huggingface: [nanochat d34](https://huggingface.co/karpathy/nanochat-d34).

 It was pretrained like this:

```bash
torchrun --standalone --nproc_per_node=8 -m scripts.base_train -- --depth=34 --device_batch_size=4 --target_param_data_ratio=40 --save_every=5000 --run=d34
```

On an 8XH100 node, which ran for \~100 hours (\~4 days) and cost \~$2,500. Notably, note that it is 2X longtrained compared to Chinchilla, i.e. the param:token ratio is overridden from 20 up to 40. This means the model was trained longer that is compute optimal, just to squeeze a bit more capability into a bit smaller package.

Some of the notable stats of the model are as follows:

```
- depth: 34
- max_seq_len: 2048
- target_param_data_ratio: 40
- device_batch_size: 4
- total_batch_size: 524,288
- Number of parameters: 2,217,082,880
- Number of FLOPs per token: 1.426509e+10
- Calculated number of iterations: 169,150
- Number of training tokens: 88,683,315,200
- Tokens : Params ratio: 40.0000
- DDP world size: 8
- Minimum validation bpb: 0.7045
- Final validation bpb: 0.7045
- MFU %: 47.40%
- Total training flops: 1.265075e+21
- Peak memory usage: 69811.85MiB
```

The achieved **CORE score** of this base model is **0.3382**. (For comparison, d32 is 0.3168, d20 "speedrun" is 0.22, GPT-2 is 0.25).

The wandb logs look like this (d32 = brown, d34 = blue, 5% resumed training = green at the very end)

<img width="1284" height="308" alt="image" src="https://github.com/user-attachments/assets/e70dc18a-a4ba-4ddf-b96c-ef844fb8537d" />

Notice that:

- d34 training was fairly stable except for a tiny bump that recovered quickly near the beginning.
- the training crashed for the last 5% with some cryptic NCCL error. So I used the new optimization resumption logic to resume training from that point for the last 5%, which worked perfect.

Midtraining and SFT then look like this:

```
| Metric          | BASE     | MID      | SFT      | RL       |
|-----------------|----------|----------|----------|----------|
| CORE            | 0.3382   | -        | -        | -        |
| ARC-Challenge   | -        | 0.5333   | 0.5435   | -        |
| ARC-Easy        | -        | 0.7071   | 0.7365   | -        |
| GSM8K           | -        | 0.1137   | 0.1296   | -        |
| HumanEval       | -        | 0.1159   | 0.0488   | -        |
| MMLU            | -        | 0.4297   | 0.4311   | -        |
| ChatCORE        | -        | 0.4087   | 0.4100   | -        |
```

For comparison, d32 ($1000 run) looked like this:

```
| Metric          | BASE     | MID      | SFT      | RL       |
|-----------------|----------|----------|----------|----------|
| CORE            | 0.3168   | -        | -        | -        |
| ARC-Challenge   | -        | 0.4787   | 0.4991   | -        |
| ARC-Easy        | -        | 0.6233   | 0.6797   | -        |
| GSM8K           | -        | 0.1099   | 0.1274   | 0.1994   |
| HumanEval       | -        | 0.1098   | 0.1280   | -        |
| MMLU            | -        | 0.3896   | 0.4049   | -        |
| ChatCORE        | -        | 0.2417   | 0.2734   | -        |
```

So our ChatCORE lifted quite a lot, from 0.27 -> 0.41. You'll notice that HumanEval stubbornly plunged during SFT, which I attribute mostly to noise. I expect one should be able to tune things a bit to recover good performance. I expected a bit more of a lift on GSM8K too - midtraining shows good lift, but SFT botches it a bit again. Possibly not tuned super well.

The d34 model is now deployed live on [nanochat.karpathy.ai](https://nanochat.karpathy.ai/)

<img width="1203" height="604" alt="image" src="https://github.com/user-attachments/assets/28e5f93e-5db9-4c86-8068-6386b6207c82" />

**Comment by nurcunal** [comment]
Will you be preparing a YouTube video that explains how to navigate the nanochat repo and implement it from scratch for enthusiasts?

**Reply by nurcunal** [reply]
Looks like we're in luck @karpathy (:
![image](https://github.com/user-attachments/assets/3bb1ddfc-718d-4571-96f0-7c34825ba9ba)

**Reply by fooSynaptic** [reply]
so for now these video uploaded already?

**Comment by mona-chen** [comment]
This is impressive Karphaty alot of this terms probably still sounds foreign to me but i sure will catch up soon following you closely xD;

**Comment by schrodingers-cat-T10** [comment]
with a lot to learn ! @karpathy did you implemented liquid neural network ?

**Comment by kenfus** [comment]
The small model smell is strong; very fun to chat with it. It's really like a small child, which always loses focus and often only mirrors back what you have typed to it.

**Comment by amjadmajid** [comment]
As always, great work! Thank you for your time and effort! 
I've interacted with nanochat, and to me, it seems to be overfitting the data (see the image below). Also, the selection algorithm might be too greedy. But very impressive and thought-provocative work. 

<img width="1404" height="1363" alt="Screenshot 2025-11-21 at 12 22 36" src="https://github.com/user-attachments/assets/ecfbe701-a4f6-4e04-b7fb-f0f2777460cb" />

**Reply by bebosudo** [reply]
same here
<img width="720" height="1017" alt="image" src="https://github.com/user-attachments/assets/6b60c8d6-deb8-45ba-a57f-78d7fcac688a" />

**Reply by Qwertyemma** [reply]
Yeah 👍, it seems like it overfitted to the data, but it’s still great and understands the user context

**Comment by john2408** [comment]
Amazing work ! Thanks for sharing this knowledge with the world. I've been working on fully understanding the implementation.

**Comment by pankajarm** [comment]
Thanks again @karpathy for all the amazing things you do for community. 

Here is [Mid Train + SFT version](https://huggingface.co/pankajmathur/nanochat-d34-finetuned) using the base mentioned above.

Dataset is same [smol-smoltalk](https://huggingface.co/datasets/HuggingFaceTB/smol-smoltalk), which is mentioned in Karpathy [OG nanochat discussion](https://github.com/karpathy/nanochat/discussions/1)

Metrics Report:
```
| Metric          | BASE     | MID      | SFT      | RL       |
|-----------------|----------|----------|----------|----------|
| ARC-Challenge   | -        | 0.5367   | 0.5418   | -        |
| ARC-Easy        | -        | 0.6961   | 0.7210   | -        |
| GSM8K           | -        | 0.1137   | 0.1327   | -        |
| HumanEval       | -        | 0.1098   | 0.1037   | -        |
| MMLU            | -        | 0.4229   | 0.4304   | -        |
| ChatCORE        | -        | 0.4045   | 0.4157   | -        |
```

**Reply by frank646745** [reply]
I was looking for something like this, how can I run it on my laptop.

**Reply by godhunter98** [reply]
I think to inference it, you can do the same via the transformers ecosystem of hugging face.

**Comment by pankajarm** [comment]
Here are all checkpoints for [RL with GSMK8 on Mid+SFT version](https://huggingface.co/pankajmathur/nanochat-d34-rl-all-ckpts)

Metrics Report:

| Metric | MID | SFT | RL | Change (SFT→RL) |
|-----------------|--------|--------|----------|-----------------|
| **GSM8K** | 0.1137 | 0.1327 | **0.2305** | **+73.6%**  |
| ARC-Easy | 0.6961 | 0.7210 | 0.7130 | -1.1% |
| ARC-Challenge | 0.5367 | 0.5418 | 0.5375 | -0.8% |
| MMLU | 0.4229 | 0.4304 | 0.4256 | -1.1% |
| HumanEval | 0.1098 | 0.1037 | 0.0671 | -35.3% |
| SpellingBee | - | - | 0.9922 | N/A |
| **ChatCORE** | 0.4045 | 0.4157 | **0.4208** | **+1.2%** |

**Comment by dcs-scd** [comment]
The nanochat architecture has changed. Are there new weights available for d34?

**Reply by jonready** [reply]
Please correct me if I’m missing something but the main change has been the removal of mid training. The base model can still be fine tuned. I’m in the middle of a training run right now on d34 base and it’s progressing really nicely on the evals with just a ton of sft.

**Comment by dcs-scd** [comment]
Ah, did not realize that  weights were so similar between the two phases.
